File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include<iostream>
#include<stack>

using namespace std;

bool solution(string s)
{
    bool answer = true;
    stack<char> st;
    for(int i=0; i<s.length(); i++){
        if(s[i]==')'){
            if(!st.empty()) st.pop();
            else return false;  
        }
        else st.push(s[i]);
    }
    if(!st.empty()) return false;

    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            if(!stack.isEmpty()){
                if(c == ')' && stack.peek() == '(') stack.pop();
                else if(c == '(') stack.push(c);
            }
            else if(c=='(') stack.push(c);
            else return false;
        }
        
        return stack.isEmpty() ? true : false;
    }
}